frame/election-provider-support/src/onchain.rs

```diff
@@ -55,77 +55,109 @@ impl From<sp_npos_elections::Error> for Error {
+
+	Ok(to_supports(&staked))
+}
+
+impl<T: Config> ElectionProvider for UnboundedOnchainExecution<T> {
+	type AccountId = T::AccountId;
+	type BlockNumber = T::BlockNumber;
+	type Error = Error;
+	type DataProvider = T::DataProvider;
+
+	fn elect() -> Result<Supports<T::AccountId>, Self::Error> {
+		elect_with::<T>(None, None)
+	}
+}
+
+impl<T: Config> InstantElectionProvider for UnboundedOnchainExecution<T> {
+	fn instant_elect(
 		maybe_max_voters: Option<usize>,
 		maybe_max_targets: Option<usize>,
-	) -> Result<Supports<T::AccountId>, Error> {
-		let voters = <Self as ElectionProvider>::DataProvider::voters(maybe_max_voters)
-			.map_err(Error::DataProvider)?;
-		let targets = <Self as ElectionProvider>::DataProvider::targets(maybe_max_targets)
-			.map_err(Error::DataProvider)?;
-		let desired_targets = <Self as ElectionProvider>::DataProvider::desired_targets()
-			.map_err(Error::DataProvider)?;
-
-		let stake_map: BTreeMap<T::AccountId, VoteWeight> = voters
-			.iter()
-			.map(|(validator, vote_weight, _)| (validator.clone(), *vote_weight))
-			.collect();
-
-		let stake_of =
-			|w: &T::AccountId| -> VoteWeight { stake_map.get(w).cloned().unwrap_or_default() };
-
-		let ElectionResult::<_, T::Accuracy> { winners: _, assignments } =
-			seq_phragmen(desired_targets as usize, targets, voters, None).map_err(Error::from)?;
-
-		let staked = assignment_ratio_to_staked_normalized(assignments, &stake_of)?;
-
-		let weight = T::BlockWeights::get().max_block;
-		frame_system::Pallet::<T>::register_extra_weight_unchecked(
-			weight,
-			DispatchClass::Mandatory,
-		);
-
-		Ok(to_supports(&staked))
+	) -> Result<Supports<Self::AccountId>, Self::Error> {
+		elect_with::<T>(maybe_max_voters, maybe_max_targets)
 	}
 }
 
-impl<T: Config> ElectionProvider for OnChainSequentialPhragmen<T> {
+impl<T: Config, VotersBound: Get<u32>, TargetsBound: Get<u32>> ElectionProvider
+	for BoundedOnchainExecution<T, VotersBound, TargetsBound>
+{
 	type AccountId = T::AccountId;
 	type BlockNumber = T::BlockNumber;
 	type Error = Error;
 	type DataProvider = T::DataProvider;
 
 	fn elect() -> Result<Supports<T::AccountId>, Self::Error> {
-		Self::elect_with(None, None)
+		elect_with::<T>(
+			Some(VotersBound::get().try_into().unwrap_or(0)),
+			Some(TargetsBound::get().try_into().unwrap_or(0)),
+		)
 	}
 }
 
-impl<T: Config> InstantElectionProvider for OnChainSequentialPhragmen<T> {
+impl<T: Config, VotersBound: Get<u32>, TargetsBound: Get<u32>> InstantElectionProvider
+	for BoundedOnchainExecution<T, VotersBound, TargetsBound>
+{
 	fn instant_elect(
 		maybe_max_voters: Option<usize>,
 		maybe_max_targets: Option<usize>,
 	) -> Result<Supports<Self::AccountId>, Self::Error> {
-		Self::elect_with(maybe_max_voters, maybe_max_targets)
+		elect_with::<T>(
+			Some(maybe_max_voters.unwrap_or(0).max(VotersBound::get().try_into().unwrap_or(0))),
+			Some(maybe_max_targets.unwrap_or(0).max(TargetsBound::get().try_into().unwrap_or(0))),
```

Comment:
```suggestion
			Some(maybe_max_voters.unwrap_or_else(Bounded::max_value).min(VotersBound::get().try_into().unwrap_or(0))),
			Some(maybe_max_targets.unwrap_or_else(Bounded::max_value).min(TargetsBound::get().try_into().unwrap_or(0))),
```

This is a little bit confusing now, but I can't think of a better way to abstract it: 

The `instantElectionProvider` has arguments that give `maybe_max_voters` and `maybe_max_targets` at run-time, on the fly. Then, they type implementing it might also have some bounds itself (e.g. if it is `BoundedOnchainExecution`). Here, you should: 

1. Interpret `maybe_max_voters` as '_there is no bound, go nuts_', ergo `.unwrap_or_else(Bounded::max_value)`
2. take the `.min` of this and the static type bounds, and chose whichever is smaller. 

How `BoundedOnchainExecution` implements `BoundedOnchainExecution` should be clearly documented in its rustdoc as well.

Reply:
What is the real difference between `elect` and `instant_elect`? Is it just the bounds? Because if so, this can now be rewritten differently.

Reply:
Honestly, I myself am a bit confused now about the exitential purpose of `InstantElect` :)) It is meant to convey two differences: 

1. Only types implement it that can instantaneously do an election (e.g. `election-provider-multi-phase::Pallet` won't implement this).
2. Additionally, just to meet the requirements of the given `governance_fallback` back then, I decided to add it to the API that instant elections can take their bounds as runtime arguments. 

But now we have this new clash: a type that wants to implement `InstantElection` might have itself some bounds on the number of voters and targets. In some sense, I am tempted to actually remove `InstantElectionProvider` since now these bounded types could work good enough. But the downside is that the ability to provide arguments via a transaction in `fn  governance_fallback` would be lost.

I still think for now my suggestion is fine, but I am curious if you have a better idea: 

> 
>   1. Interpret maybe_max_voters as 'there is no bound, go nuts', ergo .unwrap_or_else(Bounded::max_value)
>   2. take the .min of this and the static type bounds, and chose whichever is smaller.

Reply:
`InstantElectionProvider` could be removed, the interface for `ElectionProvider` could be augmented to include both methods, and rename `instant_elect` to `elect_with_bounds`.

`elect` if used with a `BoundedExecutionConfig` would be bounded, and unbounded otherwise.

`elect_with_bounds`, instead of taking an `Option`, would take the `usize` argument directly, if this is used with an `ExecutionConfig` (ie unbounded), the arguments would be the bound, and if used with `BoundedExecutionConfig`, the bound would be the minimum of the generic bounds and the method arguments.

If someone really needed the unbounded version, then it would be used with `elect` and `ExecutionConfig`, is there really a benefit to have an override where someone using `BoundedExecutionConfig` needs to remove the bounds? Is it not always better to have some sort of bounds to make sure it does not crash?

Reply:
We are mostly aligned here. Although I have to take back my suggestion of removing `InstantElectionProvider`. The reason for this is that `election_provider_multi_phase` cannot then trivially implement it, because this pallet fetches the data from the data provider before `fn elect_with_bounds()` at which point we don't know what bounds is going to be passed into `elect_with_bounds`. 

I summarize what I think is best to do as follows: 


```rust 
trait ElectionProvider {
  /// Elect, without specifying any bounds on the amount of data fetched from [`Self::DataProvider`]. An implementation could nonetheless impose its own custom limits.  
  fn elect() -> Result<_, _>;
}
 
/// Emphasize that this trait can enforce bounds on the amount of data upon call to `elect_with_bounds()`, unlike `elect` which cannot enforce this.
trait InstantElectionProvider  {
  /// Elect, ensuring that the amount of data fetched form the [`Self::DataProvider`] does not exceed  to not exceed the given bounds. An implementing type, if itself bounded, should chose the minimum of the two bounds to chose the final value of `max_voters` and `max_targets`. 
  fn elect_with_bounds(max_voters: usize, max_targets: usize)

}
```

Reply:
Shouldn't `InstantElectionProvider` be named `BoundedElectionProvider` instead?